clang/docs/SYCLFERFCIntegrationHeaderFooter.md

```diff
@@ -0,0 +1,141 @@
+
+```
+namespace sycl {
+  inline namespace _V1 {
+    namespace detail {
+      // names of all kernels defined in the corresponding source
+      static constexpr
+      const char* const kernel_names[] = {
+        "unique_name_for_first_kernel"      // first_kernel
+      };
+      // array representing signatures of all kernels defined in the
+      // corresponding source
+      static constexpr
+      const kernel_param_desc_t kernel_signatures[] = {
+        //--- first_kernel
+        { kernel_param_kind_t::kind_std_layout, 4, 0 },
+        { kernel_param_kind_t::kind_std_layout, 8, 4 },
+        { kernel_param_kind_t::kind_accessor, 4062, 12 },
+        { kernel_param_kind_t::kind_sampler, 8, 24 },
+
+        { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
+      };
+    } // namespace detail
+  } // namespace _V1
+} // namespace sycl
+```
+
+For each kernel, a mangled name associated with it, is stored in the kernel_names array.
+
+And for each kernel, each of its parameter is described in the kernel_signatures array.   The description consists of the following parts:
+
+* an encoding of the type that is captured
+* its size and its offset
+
+In the above example, the first two parameters are standard layout types (the int 'i' and the struct 'test_s'), the latter two though standard layout types are types which need special handling - accessor and sampler respectively.   The last entry acts as a terminating entry.
+
+This allows the run-time library to recreate the function object on the device.
+
+The full set of encoding is:
+
+* kind_accessor  for sycl accessor types
+* kind_std_layout for standard layout types
+* kind_sampler for sycl sampler types
+* kind_pointer for pointer types
+* kind_specialization_constants_buffer for initializing specialization constants
+* kind_stream for sycl stream types and
+* kind_invalid used to indicate the terminal value in the signatures array.
+
+
+# Integration Footer
+An integration footer is a source file produced during device compilation that is intended to be post-included during host compilation of the same source file.
+
+The SYCL 2020 Specification defines a specialization constant as a constant variable where the value is not known until compilation of the SYCL kernel function.
+
+Consider the following example:
+```
+  constexpr specialiation_id<int> int_const;
+  class Wrapper {
+  public:
+    static constexpr specialization_id<float> float_const;
+  };
+```
+
+Also note that the SYCL header files declare, but not define, a special function to obtain the numeric id of a specialization constant.
+```
+     namespace detail {
+         template<auto &SpecConstName>
+         inline const char *get_spec_constant_symbolic_ID();
+     }
+```
+
+In order to accommodate both implementations that have native support for this as well as those that do not, we propose the following integration footer mechanism.
```

Comment:
I think the key thing here is that the partial specializations of `get_spec_constant_symbolic_ID` are partially specialized using a reference to the variable of type `specialiation_id`.  Since these variables are defined in the user's application, the partial specializations cannot be defined in the integration header.  Instead, they are defined in the integration footer, which is included after the user's application and thus after the definitions of the `specialiation_id` variables.